### ISSUE TITLE
[Bugs]: Tinker Events Bugs

### ISSUE THREAD

**Issue by Lanse505** [SlimeKnights/TinkersConstruct#3471]
# Pre-text: 
The code that is gonna be on github is WIP-code and I'm still new to modding and coding in general so please be aware my code will probably burn holes in your retinas!

You have been warned >.>

-----------------------------------------

**Versions:**
TCon: 1.12.2-2.10.0.72
Mantle: 1.12-1.3.2.2
Forge: 1.12.2-14.23.2.2641

**Link to my Github Project:**
https://github.com/Coders-After-Dark/CompatSkills

**Link to Package in Question:**
https://github.com/Coders-After-Dark/CompatSkills/tree/master/src/main/java/codersafterdark/compatskills/common/compats/tinkersconstruct

-----------------------------------------

Here are some weird oddities I've found when trying to work with the TinkerCraftingEvent and it's sub-events. I'm fairly sure it's not my fault, but knowing me it is my fault and I'm missing something completely obvious, but hey it's worth shooting up an issue about.

- All of the events get fired a good 5-6 times per trigger.
This means everytime the GUI is opened, item stacks are changed, etc.

- Not sure if this is for all events but at least for the ToolCraftingEvent.
When it gets fired several times, the first 2 times always returns a null EntityPlayer.

**Comment by bonii-xx** [comment]
Not odd per se. You'll always have at least 2 events fired, because you're in an SSP environment where server and client are the same (you'll get the event on the server as well as the client, which are separate machines in multiplayer).

The crafting events may be fired multiple times depending on what you do because of some checks and the tool being rebuilt from scratch. That's not a problem, since your events need to be idempotent anyway.

Do you have any specific problems?

**Comment by Lanse505** [comment]
So the issue this causes at least for the ToolCraftingEvent is that since it can't cancel the first two events being fired, for the event cancellation I'm doing for the ToolCraftingEvent here: [Link](https://github.com/Coders-After-Dark/CompatSkills/blob/master/src/main/java/codersafterdark/compatskills/common/compats/tinkersconstruct/toollocks/ToolTypeLockHandler.java)

I end up with issues like this:
https://i.imgur.com/tOtbEHJ.png
https://i.imgur.com/Qd6E1XQ.png

Where in whenever you close and open the GUI, it doesn't respect the event cancellation and thus allows you to craft the tool, however, once you pull out the item it figures out it's supposed to cancel the event, creates a tool item stack but doesn't remove anything from the item stacks, creating a duped item. 

Sorry if my explanations are a bit disjointed!

**Comment by pupnewfster** [comment]
Due to some refactoring of how we are handling the events the new location of the code in question is https://github.com/Coders-After-Dark/CompatSkills/blob/master/src/main/java/codersafterdark/compatskills/common/compats/tinkersconstruct/TinkerLockHandler.java#L54

The issue we are having is we want to cancel the events based on if the player has met a certain set of requirements yet. `TinkerCraftingEvent.ToolCraftingEvent` fires with event.getPlayer() as null when the gui is initially opened. We are unable to check if the player has met the requirements as until they interact with the GUI the only event we have received is one that we cannot determine the player from.

Is there a reason that the initial firing of this event (twice on single player which is fine) does not include what player caused the event to be fired? The other TinkerCraftingEvents do fire initially with the Player object set. If for technical reasons it must initially fire as null, would it be possible to fire the event AGAIN with a player object set once the gui is done being opened as a way to get the player object before they interact with the state of the gui? (The second option is by far the more messy and less efficient solution.)

**Comment by bonii-xx** [comment]
Looks like the gui container wasn't fully initialized yet when the event was fired. Just a matter of reordering.

**Comment by bonii-xx** [comment]
For your dev environment, anything from build 2.10.0.77 on contains the fix.

**Comment by pupnewfster** [comment]
2.10.0.77 seems to partially fix the issue. It now works properly, EXCEPT that it shows a blank info panel when the gui is opened (https://gyazo.com/56d01b56acb70763eaf98331237bc57a) instead of showing the error message we set. This also happens if I change Minecraft's window size (it shows an error and then after resizing it changes to the blank info panel).

Resizing the gui does not have issues with the PartBuilder. I will try to look into this more tomorrow, I believe that the gui may not be updating for the player on open for the ToolStation.

**Comment by pupnewfster** [comment]
I am trying to figure out the fix so that I can create a pull request, but the issue is that this if statement:
https://github.com/SlimeKnights/TinkersConstruct/blob/1.12/src/main/java/slimeknights/tconstruct/tools/common/client/GuiToolStation.java#L261

is firing as true and resetting the GUI sidebar for when the gui is resized or initialized after having been cancelled.

So if you know a good way to fix that go for it, and I will keep looking into it to see if I can find a way.

**Comment by pupnewfster** [comment]
Update: I have no idea how to fix it without breaking some other things. On closer inspection though, I believe it is updateDisplay() in general in the class I linked that causes the problem as if you are trying to modify a tool it shows information about the tool when you first open the gui instead of just showing a blank gui.

I think updateDisplay() will need to be updated somehow to not actually "update" if there is an error message because of the last event being cancelled.

**Comment by bonii-xx** [comment]
That's a more general issue because there's a lot of syncing required. Out of scope of this ticket, and as long as it works probably not worth investing the work required while there's more important things to do.
